#24 Codify your Developer VMs! (#139)
Zuehlke/fifty-shades: codify-your-developer-vms.md

to work on a specific project (including compiler toolchains, IDEs, sytem wide settings, etc...).
The aim is to ensure a consistent environment not only between team members of that project (which may have
completely different operating systems on their laptops) but also with the build agents in the CI environment.

## Why would I need a Developer VM anyway?

Some people hate it, others love it. However you put it, it is one sure-fire way to ensure that the
whole development team uses a consistent environment and avoids the typical “works on my machine”
issues we all know about. Especially for us, executing and delivering a multitude of projects (and engineers
joining and leaving these projects more or less frequently), developer VMs have become an indispensable tool.
Not only for the consistency aspect, but also for reducing ramp-up times or being able to easily archive the
development environment when the project enters the retirement phase.

On regulated projects this is even more relevant, since you need to validate and verify the development toolchain
and need to ensure that everyone uses exactly that defined toolchain.

> 

Author: I understand that not everybody is a fan of running a VM in order to be able to work for a specific project, that's why I believe you can't or shouldn't force anyone to use it. In regulated projects, that might be easier as you might be forced by regulatory requirements to do that (i.e. the blame is not on you then ;)).

My observations so far:

* whenever you have Mac users, they are quite thankful to be provided with a VM if they work in a project where Windows is a requirement
* if you have people which frequently switch between projects, they are usually happy to have their toolchains / environments cleanly separated
* if you have people which are full time working for one project which has their preferred + installed operating system as a requirement, they rather work in their OS directly than in a VM
* disk space is an issue with developer VMs, that's definitely a cause for annoyance ;)

> 

Author: not sure though how I could write this in a more convincing style (and it won't help the haters anyway, probably ;)). 

@abeggchr if you have specific ideas on how to improve this I would happily take them though :)

> 

Reviewer: If I had, I would have convinced more people to use a VM in our project.

## OK, but why should I automate it?

While we started out to craft these VMs manually (and with love) in the early days, we can nowadays easily

> automate the whole procedure with tools like Vagrant, Chef, Ansible and co. Applying automation and

Reviewer 2: "hardcore .Netties" should probably be "old MS gits nearing retirement" and they need to get up with the times then :) These all work with Windows and have done for a while :)

> automate the whole procedure with tools like Vagrant, Chef, Ansible and co. Applying automation and

Author: Indeed that may sound like Linux only, but all of these tools do support Windows as well. By design they are actually OS agnostic. Will resolve that comment

Infrastructure-as-Code principles to developer VMs not only increases transparency and reproducibility,
but also allows for an in-VM update mechanism to maintain consistency while the toolchain evolves during
the lifetime of the project.

Finally, an automated developer VM should yield the following properties:

* updatability - being able to update an existing VM with the push of a button from within the VM
* testability - running automated tests in the VM to ensure that it works as expected
* installability - the VM should distributed as a standard VirtualBox / VMware image
* adaptability - developers should be able to easily adapt and improve the VM as the toolchain evolves
* reproducibility - changes to the development environment are transparent, traceable and reproducible

> 

Reviewer: Gut und recht, aber der Versuch, die DevVM zu automatisieren ist in unserem Projekt grandios gescheitert. Es gab Probleme bei der MSSQL Installation (musste manuell ausgeführt werden weil nicht mit Chocolatey möglich), Office kann nicht installiert werden (weil die einen Teammembers eine Office 365 Lizenz haben, die anderen Office 2019 über MSDN), Windows an sich (schonmal eine Herausforderung mit der Lizenz) und dann und dann gibt es noch 1000 (oder sagen wir mal eine handvoll) Einstellungen welche doch noch manuell ausgeführt werden müssen: vs code plugins installieren, visual studio immer als admin ausführen, ...

Okay, genug Frust abgeladen. Den Artikel oben ist gut so und passt auch auf dem Artikel Level. Meine ganzen Umsetzungsfragen müssen ja da nicht beantwortet werden.

Reviewer 2: Und fuer jeden der aufgelistet Probleme eine Liste. OK, ich habe gesagt ich werde Torbens Artikel reviewen, nicht die Kommentare dazu...

## How can I automate it?

The good news is that there are quite a few so-called "Configuration Management" tools out there, which essentially
allow you to automate the provisioning of software and configuration on top of an existing operating system
(in our case a fresh installation of the operating system of choice, within a VM image). 

The bad news is that you have to choose one -- and in addition to the configuration management tools that do the
provisioning, there is (gladly) also a multitude of tools for testing the configuration of such a (virtual) machine.

> Assuming you have a preference for either Ruby or Python, these are the tools you will likely end up with:

Reviewer: ```suggestion
Depending on your preference, these are the tools you will likely end up with:
```

* Rubyists will like
  * Chef - as the configuration management tool
  * Foodcritic - for linting their Chef recipes
  * InSpec / ServerSpec - for testing the Chef-provisioned systems

* Pythonistas probably prefer
  * Ansible - as the configuration management tool
  * ansible-lint - for linting their Ansible roles
  * TestInfra - for testing the Ansible-provisioned systems

## I’m sold, sounds like pretty cool stuff! What is the fastest way to get there?

It may sound a bit of effort to achieve this if you start from scratch, so that’s why we have a RUA with a

> template / skeleton project along with a tutorial for you. No excuses for doing it manually anymore! ;-)

Reviewer: ```suggestion
template / skeleton project along with a tutorial for you.
```

Reviewer 2: Not everyone will be familiar with the RUA abbreviation

Author: @damphyr good point, I removed the reference to RUA

@abeggchr @damphyr I was a bit unsure about this last paragraph in general, considering that the article's content / text is there to stay, while the examples might be broken / non-functional at some point in time. Still they are useful. At least I reduced them to stuff that is hosted on our zuehlke Github account (which is fair to promote I guess).

Still I'm not 100% convinced about referencing these. What's your take on this?
